src/gurt/telemetry.c

```diff
@@ -1824,6 +1853,269 @@ d_tm_add_metric(struct d_tm_node_t **node, char *metric, int metric_type,
+ *				node.
+ * \param[in]	num_buckets	Specifies the number of buckets the histogram
+ *				should have.
+ * \param[in]	initial_width	The number of elements in the first bucket
+ * \param[in]	multiplier	Applied to the second .. nth bucket to
+ *				increase the width of each successive bucket
+ *				by this factor over the previous bucket.
+ *				A multiplier of 1 creates equal size buckets.
+ *				A multiplier of 2 creates buckets that are
+ *				twice the size of the previous bucket.
+ *
+ * \return			D_TM_SUCCESS		Success
+ *				-DER_INVAL		node, path, num_buckets,
+ *							initial_width or
+ *							multiplier is invalid.
+ *				-DER_OP_NOT_PERMITTED	Node was not a gauge
+ *							or duration.
+ *				-DER_NO_SHMEM		Out of shared memory
+ *				-DER_NOMEM		Out of heap
+ */
+int
+d_tm_init_histogram(struct d_tm_node_t *node, char *path, int num_buckets,
+		    int initial_width, int multiplier)
+{
+	struct d_tm_metric_t	*metric;
+	struct d_tm_bucket_t	*dth_buckets;
+	uint64_t		min = 0;
+	uint64_t		max = 0;
+	int			rc = D_TM_SUCCESS;
+	char			*meta_data;
+	char			*fullpath;
+	int			i;
+
+	if (node == NULL)
+		return -DER_INVAL;
+
+	if (path == NULL)
+		return -DER_INVAL;
+
+	if (num_buckets < 2)
+		return -DER_INVAL;
+
+	if (initial_width < 1)
+		return -DER_INVAL;
+
+	if (multiplier < 1)
+		return -DER_INVAL;
+
+	if (!((node->dtn_type == D_TM_GAUGE) ||
+	      (node->dtn_type & D_TM_DURATION)))
+		return -DER_OP_NOT_PERMITTED;
+
+	rc = D_MUTEX_LOCK(&d_tm_add_lock);
+	if (rc != 0) {
+		D_ERROR("Failed to get mutex: " DF_RC "\n", DP_RC(rc));
+		goto failure;
+	}
+
+	metric = node->dtn_metric;
+
+	metric->dtm_histogram = d_tm_shmalloc(sizeof(struct d_tm_histogram_t));
+	if (metric->dtm_histogram == NULL) {
+		D_MUTEX_UNLOCK(&d_tm_add_lock);
+		rc = -DER_NO_SHMEM;
+		goto failure;
+	}
+
+	metric->dtm_histogram->dth_buckets = d_tm_shmalloc(num_buckets *
+						  sizeof(struct d_tm_bucket_t));
+	if (metric->dtm_histogram->dth_buckets == NULL) {
+		D_MUTEX_UNLOCK(&d_tm_add_lock);
+		rc = -DER_NO_SHMEM;
+		goto failure;
+	}
+	metric->dtm_histogram->dth_num_buckets = num_buckets;
+	metric->dtm_histogram->dth_initial_width = initial_width;
+	metric->dtm_histogram->dth_value_multiplier = multiplier;
+
+	D_MUTEX_UNLOCK(&d_tm_add_lock);
+
+	dth_buckets = metric->dtm_histogram->dth_buckets;
+
+	min = 0;
+	max = initial_width - 1;
+	for (i = 0; i < num_buckets; i++) {
+		D_ASPRINTF(meta_data, "histogram bucket %d [%lu .. %lu]",
+			   i, min, max);
+		if (meta_data == NULL) {
+			rc = -DER_NOMEM;
+			goto failure;
+		}
+
+		D_ASPRINTF(fullpath, "%s/bucket %d", path, i);
+		if (fullpath == NULL) {
+			rc = -DER_NOMEM;
+			goto failure;
+		}
+
+		dth_buckets[i].dtb_min = min;
+		dth_buckets[i].dtb_max = max;
+
+		rc = d_tm_add_metric(&dth_buckets[i].dtb_bucket, fullpath,
+				     D_TM_COUNTER, meta_data, "");
+		D_FREE(fullpath);
+		D_FREE(meta_data);
+		if (rc)
+			goto failure;
+
+		min = max + 1;
+
+		if (i == (num_buckets - 2)) {
+			max = UINT64_MAX;
```

Comment:
Not sure I understand what's happening here. The last bucket is huge? Why?
Would be a good place to add a comment explaining.

Reply:
I think it's because the last bucket contains all possible values (infinite range) above what is specified in the previous bucket ranges i.e. "everything else".

Reply:
Correct.  The last bucket needs to store the maximum possible value, unless we decided that we want to filter out values that don't fit into the last bucket specified.